keystore/v2/keystore/filesystem/export.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *KeyStore) importKeyRing(newRingData *asn1.KeyRing, delegate *api.KeyRingImportDelegate) error {
+	keyRing := newKeyRing(s, string(newRingData.Purpose))
+	err := s.readKeyRing(keyRing)
+	switch err {
+	case nil:
+		// If the key store successfuly returned an existing key ring with the same name,
+		// we have to resolve this conflict somehow. Present both current and new versions
+		// to the delegate and let it decide how to proceed.
+		decision := api.ImportAbort
+		err := ErrKeyRingExists
+		if delegate != nil && delegate.DecideKeyRingOverwrite != nil {
+			decision, err = delegate.DecideKeyRingOverwrite(keyRing.data, newRingData)
+		}
+		switch decision {
+		case api.ImportOverwrite:
+			// Forget whatever we just read and import into a clean key ring.
+			keyRing = newKeyRing(s, string(newRingData.Purpose))
+			err := keyRing.importASN1(newRingData)
+			if err != nil {
+				return err
+			}
+		case api.ImportSkip:
+			return nil
+		default:
+			return err
+		}
+
+	case backendAPI.ErrNotExist:
+		// If the key ring does not seem to exist right now, go ahead with clean-slate import.
+		err := keyRing.importASN1(newRingData)
+		if err != nil {
+			return err
+		}
+
+	default:
+		// Otherwise, this is some unexpected error from the key store. Abort import and get out.
+		return err
+	}
+	return s.writeKeyRing(keyRing)
+}
+
+var exportKeyContext = []byte("AKSv2 keystore: exported key rings")
+
+func (s *KeyStore) encryptAndSignKeyRings(rings []asn1.KeyRing, cryptosuite *crypto.KeyStoreSuite) ([]byte, error) {
+	keysData := &asn1.EncryptedKeys{KeyRings: rings}
+	keysBytes, err := keysData.Marshal()
+	if err != nil {
+		return nil, err
+	}
+	defer utils.FillSlice(0, keysBytes)
+
+	encryptedKeyBytes, err := cryptosuite.KeyEncryptor.Encrypt(keysBytes, exportKeyContext)
+	if err != nil {
+		return nil, err
+	}
+	container := asn1.SignedContainer{Payload: asn1.SignedPayload{
+		ContentType:  asn1.TypeEncryptedKeys,
+		Version:      asn1.KeyRingVersion2,
+		LastModified: time.Now(),
+		Data:         encryptedKeyBytes,
+	}}
+
+	notary, err := signature.NewNotary(cryptosuite.SignatureAlgorithms)
+	if err != nil {
+		return nil, err
+	}
+	signedKeyContainer, err := notary.Sign(&container, exportKeyContext)
+	if err != nil {
+		return nil, err
+	}
+	return signedKeyContainer, nil
+}
+
+func (s *KeyStore) decryptAndVerifyKeyRings(ringData []byte, cryptosuite *crypto.KeyStoreSuite) ([]asn1.KeyRing, error) {
+	panic("not implemented")
+}
+
+func (r *KeyRing) exportASN1() (exported asn1.KeyRing, err error) {
+	exported = asn1.KeyRing{
+		Purpose: r.data.Purpose,
+		Current: r.data.Current,
+		Keys:    make([]asn1.Key, len(r.data.Keys)),
+	}
+	copy(exported.Keys, r.data.Keys)
+	defer func() {
+		if err != nil {
+			zeroizeKeyRing(&exported)
+		}
+	}()
+	for i := range exported.Keys {
+		decrypted, err := r.decryptAllKeyData(exported.Keys[i].Data, exported.Keys[i].Seqnum)
```

Comment:
why we decrypt data for export? to encrypt it with "export" key and allow to encrypt with another key on import operation?

Reply:
> why we decrypt data for export? to encrypt it with "export" key and allow to encrypt with another key on import operation?

Yes, that is correct. Suppose you’re transferring keys between key stores A and B. Each of them uses their own encryption keys and at the moment of export from A we don’t know where the keys will be imported to. During export we’ll decrypt them with key A, encrypt them back with a temporary key T. During import the T key will be transferred to the B place (in some unspecified, secure way), then the keys will be decrypted with T and encrypted for B with B’s key.